---
title: "halfBaked RNA-seq Pin Generation Template"
author: 
  - name: Jared Andrews
    affiliation:
    - St. Jude Children's Research Hospital
    email: jared.andrews07@gmail.com
output: 
  md_document:
    preserve_yaml: false
date: "`r doc_date()`"
package: "`r pkg_ver('halfBaked')`"
vignette: >
  %\VignetteIndexEntry{halfBaked Pin Generation Template}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
params:
  board: "your_server_info"
  pin_name: "PinGen_Name_RNAseq_SummarizedExperiment_mm10"
  pin_description: "A SummarizedExperiment containing O4+ immunopanned oligodendrocyte precursor cells (OPCs) from cortices of P5 or P6 Eed KO or control mice."
---

```{r setup, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(error = FALSE, message = FALSE, warning = FALSE, eval = FALSE)
```

# Study - Experiment - RNA-seq Data - Pin Generation

This notebook details how to create re-usable and easily shared [pins](https://pins.rstudio.com/articles/pins.html) for RNA-seq data.
These pins will contain a `SummarizedExperiment` object containing gene counts, normalized counts, transformed counts, and TPMs along with sample metadata.
In short, pretty much everything one would need to perform downstream analyses.

For convenience, we'll also store experiment/dataset description, the code used to generate the pin (i.e. this notebook).

It's recommended to detail how gene counts were generated here.

We typically use the [nf-core RNA-seq pipeline](https://nf-co.re/rnaseq/) to generate BAMs and salmon quantifications, which are then used to generate gene counts appropriate for downstream use.
This is convenient as the samplesheets can double as our sample metadata tables here.

Noting the pipeline version, genome information, and any altered parameters here is recommended.

As an example, this notebook details pin generation for data from [GSE135880](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE135880), which was processed with the nf-core RNA-seq pipeline:

```{bash, eval=FALSE}
nextflow run nf-core/rnaseq -r 3.12.0 -profile singularity -c "$BAKER_REF"/nf_configs/rnaseq.config -w /scratch_space/jandrews/"$LSB_JOBNAME" \
--outdir ./nfcore_mm10 --email jared.andrews@stjude.org --input nfcore_rnaseq.samplesheet.csv --gencode --genome MM10 --aligner star_salmon \
--pseudo_aligner salmon --max_memory 128.GB --skip_stringtie --max_multiqc_email_size 15.MB -resume
```

## Params Usage

This notebook uses YAML [params](https://bookdown.org/yihui/rmarkdown/parameterized-reports.html) in the header to specify the pin name and board to upload the pin to.
This allows for easy re-use of the notebook for different datasets and boards in a lazy fashion.

Read more about the [pins](https://pins.rstudio.com/articles/pins.html) package if you don't know what a "board" is or how to use them.

# Experiment Info

```{r, eval = FALSE}
description <- "
  **Study - SubStudy - Experiment - RNA-seq Data**

  This dataset contains O4+ immunopanned oligodendrocyte precursor cells (OPCs) from cortices of P5 or P6 Eed KO or control mice.
  See the [associated publication](https://www.science.org/doi/10.1126/sciadv.aaz6477) for more details.

  Please see the Pin generation code to view how this data was processed.
  "
```

# Load Data & Create SummarizedExperiment

```{r load, eval = FALSE}
library(SummarizedExperiment)
library(readr)
library(pins)

# Load sample metadata.
meta <- read.csv("nfcore_rnaseq.samplesheet.csv", header = TRUE, stringsAsFactors = TRUE)

# Drop FASTQ file locations.
meta <- meta[, !colnames(meta) %in% c("fastq_1", "fastq_2")]

# Load counts. This object was generated using tximport via the nf-core RNA-seq pipeline on the salmon quants and
# is appropriate for pretty much all downstream DE packages (DESeq2, edgeR, limma).
cts <- read.table("salmon.merged.gene_counts_length_scaled.tsv", header = TRUE, sep = "\t", stringsAsFactors = FALSE)

# nf-core RNA-seq pipeline generates a counts table with the first two columns as gene IDs and gene symbols.
genes <- cts[, 1:2]
rownames(cts) <- cts[, 1]
cts <- cts[, -c(1:2)]

# Also carry along TPMs as an additional assay.
tpms <- read.table("salmon.merged.gene_tpm.tsv", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
rownames(tpms) <- tpms[, 1]
tpms <- tpms[, -c(1:2)]

# Set metadata rownames and ensure they match count column names.
rownames(meta) <- meta$sample
all(rownames(meta) == colnames(cts))

# Create a SummarizedExperiment object.
se <- SummarizedExperiment(
    assays = list(counts = as.matrix(cts),
                  tpm = as.matrix(tpms),
                  log2tpm = log2(as.matrix(tpms) + 1)),
    colData = meta,
    rowData = genes
)
```

## Add Metadata to Summarized Experiment and Upload Pin

```{r, eval = FALSE}
# Render this notebook and add it to the metadata of the pin along with experiment metadata.
pin_code <- read_file(rmarkdown::render("0.RNAseq_PinGen_Template.Rmd", quiet = TRUE))
metadata(se) <- list(pin_code = pin_code, description = description)

# Save the object locally as well.
saveRDS(se, file = paste0(params$pin_name, ".rds"))

# Here you would connect to your board and write the pin to it.
board <- board_connect(server = params$board)
pin_write(board, se, type = "rds", name = params$pin_name, title = params$pin_name, description = params$pin_description)
```

# Session Info

```{r sessinfo, eval = TRUE}
sessionInfo()
```

# References
